File: LC/[0583] Delete Operation for Two Strings - dp.cpp
class Solution {
    public:
        int minDistance(string word1, string word2) {
            int N1=word1.size();
            int N2=word2.size();
    
            string ss = "#"+word1;
            string tt = "#"+word2;
    
            //declare and define dp
            /*
            dp[ii][jj]: min delete required to make ss[:ii+1],  t[:j+1] the same
            */
            auto dp = vector<vector<int>>(N1+1, vector<int>(N2+1, INT_MAX/2));
            
            //initialize dp
            for(int ii=1; ii<=N1; ++ii)dp[ii][0]=ii;
            for(int jj=1; jj<=N2; ++jj)dp[0][jj]=jj;
            dp[0][0]=0;
    
            for(int ii=1; ii<=N1; ++ii){
                for(int jj=1; jj<=N2; ++jj){
                    if(ss[ii]==tt[jj]){
                        dp[ii][jj]=dp[ii-1][jj-1];
                    }else{
                        dp[ii][jj] = min(dp[ii-1][jj], dp[ii][jj-1])+1;
                    }
                }
            }
    
    
            return dp[N1][N2];
        }
    };

File: LC/[0583] Delete Operation for Two Strings - dp.java
class Solution {
    public int minDistance(String word1, String word2) {
        int N1=word1.length();
        int N2=word2.length();
        String ss = "#"+word1;
        String tt = "#"+word2;

        //declare dp
        var dp = IntStream.range(0, N1+1)
                 .mapToObj(ii->IntStream.range(0, N2+1)
                               .map(jj->Integer.MAX_VALUE/2)
                               .toArray()
                          ).toArray(int[][]::new);
        ;

        //initialize dp
        //if ss is null
        for(int jj=1; jj<=N2; ++jj) dp[0][jj]=jj;
        //if tt is null
        for(int ii=1; ii<=N1; ++ii) dp[ii][0]=ii;

        dp[0][0]=0;

        //populate dp
        for(int ii=1; ii<=N1; ++ii){
            for(int jj=1; jj<=N2; ++jj){
                if(ss.charAt(ii)==tt.charAt(jj)){
                    dp[ii][jj] = Math.min(dp[ii][jj], dp[ii-1][jj-1]);
                }else{
                    dp[ii][jj] = Math.min(dp[ii][jj], Math.min(dp[ii-1][jj], dp[ii][jj-1])+1);
                }
            }
        }

        return dp[N1][N2];
    }
}